File: solution/1669. Merge In Between Linked Lists/1669.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeInBetween(ListNode* list1, int a, int b, ListNode* list2) {
        auto p = list1, A = list1, B = list1;
        for (int i = 0; i <= b; ++i, p = p->next) {
            if (i == a - 1)
                A = p;
            if (i == b)
                B = p->next;
        }
        A->next = list2;
        
        // find tail of list2
        p = list2;
        while (p->next) {
            p = p->next;
        }
        p->next = B;
        
        return list1;
    }
};

// 0 1  2 3 4 5  6 7 8 9 ...
// x x [x x x x] x x x x ...
//   A  a     b  B

File: solution/1669. Merge In Between Linked Lists/1669.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeInBetween(self, list1: ListNode, a: int, b: int, list2: ListNode) -> ListNode:
        p = A = B = list1
        
        for i in range(b + 1):
            if i == a - 1:
                A = p
            if i == b:
                B = p.next
            p = p.next
        A.next = list2

        p = list2
        while p.next:
            p = p.next
        p.next = B

        return list1
